File: 0229-majority-element-ii/0229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int cnt1=0, cnt2=0, el1=0, el2=0;
        int n = nums.size();
        vector<int> ans;
        int mini = (int)(n/3)+1;
        if(n==1) return nums;
        for(int i=0;i<n;i++){
            if(cnt1==0 && el2!=nums[i]){
                cnt1=1;
                el1=nums[i];
            }
            else if(cnt2==0 && el1!=nums[i]){
                cnt2=1;
                el2=nums[i];
            }
            else if(el1==nums[i]) cnt1++;
            else if(el2==nums[i]) cnt2++;
            else{
                cnt1--, cnt2--;
            }

        }
        cnt1=0;
        cnt2=0;
        for(int i=0;i<n;i++){
            if(el1==nums[i]) cnt1++;
            else if(el2==nums[i]) cnt2++;
        }
        
        if(cnt1>=mini){
            ans.push_back(el1);
        }
        if(cnt2>=mini){
            ans.push_back(el2);
        }
        return ans;
        
    }
};

File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> ans = new ArrayList<>();
      HashMap<Integer, Integer> mp = new HashMap <>();
      int n = nums.length;
      for(int i=0; i<n; i++){
        if(mp.containsKey(nums[i])){
            mp.put(nums[i], mp.get(nums[i])+1);
        }
        else{
            mp.put(nums[i], 1);
        }
      }
      for(int key: mp.keySet()){
        if(mp.get(key)>n/3){
            ans.add(key);

        }
      }  
      return ans;
    }
}